### PR TITLE
feat(Windows): #2998, add 'signing in' state to system tray menu

### LINKED ISSUE #2998: windows: show "Signing in..." in tray menu during auth flow
The sign-in flow should time out after a certain time. Trying to initiate a second sign-in should be ignored as invalid? Or cancel the first?

Should also apply if we reload a token and sign in without user clicking "Sign in"

### PR DESCRIPTION
**Pull request by ReactorScram**
Depends on the new auth flow in #3147 

I couldn't find the UI mockups so I just modified the "Sign In" menu to create this.

### CODE REVIEW

File: rust/windows-client/src-tauri/src/client/gui/system_tray_menu.rs [hunk 1]
```diff
@@ -66,6 +66,16 @@ pub(crate) fn signed_in(user_name: &str, resources: &[ResourceDescription]) -> S
     menu
 }
 
+pub(crate) fn signing_in() -> SystemTrayMenu {
+    // TODO: Check this layout with Jamil, I lost track of the Figma for it.
```
Comment (jamilbk) [inline rust/windows-client/src-tauri/src/client/gui/system_tray_menu.rs:70]: Post a screenshot? I'm not sure if we had one for "signing in" state

Comment (ReactorScram) [reply rust/windows-client/src-tauri/src/client/gui/system_tray_menu.rs:70]: image

Comment (jamilbk) [reply rust/windows-client/src-tauri/src/client/gui/system_tray_menu.rs:70]: Lgtm

File: rust/windows-client/src-tauri/src/client/gui.rs [hunk 2]
```diff
@@ -411,20 +411,39 @@ impl Controller {
...
-            .auth
-            .session()
-            .ok_or_else(|| anyhow!("connlib is signed in but auth module isn't"))?
-            .actor_name;
-        self.app
-            .tray_handle()
-            .set_menu(system_tray_menu::signed_in(actor_name, &resources))?;
+    /// Returns a new system tray menu
+    fn build_system_tray_menu(&self) -> Result<tauri::SystemTrayMenu> {
+        Ok(if let Some(auth_session) = self.auth.session() {
```
Comment (conectado) [inline rust/windows-client/src-tauri/src/client/gui.rs]: I'd rather we have a separate variable that we warp in `Ok`, this looks a bit weird

File: rust/windows-client/src-tauri/src/client/gui.rs [hunk 3]
```diff
@@ -411,20 +411,39 @@ impl Controller {
...
+            if let Some(connlib_session) = &self.session {
+                // TODO: Modify CallbackHandler so it can tell us whether the tunnel is really ready.
+                let tunnel_ready = true;
+                if tunnel_ready {
```
Comment (conectado) [inline rust/windows-client/src-tauri/src/client/gui.rs]: we can leave the todo but I'd rather remove this `if true`

File: rust/windows-client/src-tauri/src/client/gui.rs (hunk 2 above)
Comment (conectado) [reply rust/windows-client/src-tauri/src/client/gui.rs]: We should just create some helper functions to reduce nesting here, like `signing_in()`, `signed_in()` and `signed_out()` to reduce nesting, it's kind of hard to parse as is.

File: rust/windows-client/src-tauri/src/client/gui.rs [hunk 4]
```diff
@@ -411,20 +411,39 @@ impl Controller {
...
+            system_tray_menu::signed_out()
+        })
+    }
+
+    /// Builds a new system tray menu and applies it to the app
+    fn refresh_system_tray_menu(&self) -> Result<()> {
+        let menu = self.build_system_tray_menu().unwrap_or_else(|error| {
+            tracing::error!(?error, "couldn't build system tray menu");
+            // If there's an error, show the signed-out menu as a default.
+            system_tray_menu::signed_out()
```
Comment (conectado) [inline rust/windows-client/src-tauri/src/client/gui.rs]: if we return `signed_out` by default in `build_system_tray_menu` we don't need the additional flow control here.

File: rust/windows-client/src-tauri/src/client/gui.rs [hunk 5]
```diff
@@ -411,20 +411,39 @@ impl Controller {
...
+                if tunnel_ready {
+                    // Signed in, tunnel ready
+                    let resources = connlib_session.callback_handler.resources.load();
+                    system_tray_menu::signed_in(&auth_session.actor_name, &resources)
+                } else {
+                    // Signed in, raising tunnel
+                    system_tray_menu::signing_in()
+                }
+            } else {
+                bail!("We have an auth session but no connlib session");
```
Comment (conectado) [inline rust/windows-client/src-tauri/src/client/gui.rs]: I think this should be a `debug_assert`

**Review by conectado — APPROVED** [review]
Left some style comments but looks good!

File: rust/windows-client/src-tauri/src/client/gui.rs (hunk 3 above)
Comment (ReactorScram) [reply rust/windows-client/src-tauri/src/client/gui.rs]: I wanted to make space for it in the control flow. Luckily the PR for this feature landed already so I just implemented that, too, so it'll say "Signing in..." after auth is done and before connlib gives us `on_tunnel_ready`

File: rust/windows-client/src-tauri/src/client/gui.rs (hunk 5 above)
Comment (ReactorScram) [reply rust/windows-client/src-tauri/src/client/gui.rs]: I made it a `tracing::error`, I want it to be in release logs since it's a logic error that's cheap to find, but yeah maybe it shouldn't panic in release.

File: rust/windows-client/src-tauri/src/client/gui.rs (hunk 2 above)
Comment (ReactorScram) [reply rust/windows-client/src-tauri/src/client/gui.rs]: I tried but in the end just left another TODO. 
There are at least 3 different pieces of state to query to figure out whether we're signed in or not:
- Whether a connlib session exists
- Whether we got tunnel_ready, which AFAICT the connlib session doesn't expose directly
- What state the auth state machine is in, since the app as a whole is still signed in even if the user somehow managed to start a second auth flow right as the first one was completing

I can probably refactor the auth state machine later to simplify some of that.
